azure-iot-device/tests/iothub/aio/test_async_handler_manager.py

```diff
@@ -0,0 +1,264 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import logging
+import pytest
+import asyncio
+import inspect
+from azure.iot.device.iothub.aio.async_handler_manager import AsyncHandlerManager
+from azure.iot.device.iothub.aio.async_handler_manager import MESSAGE, METHOD, TWIN_DP_PATCH
+from azure.iot.device.iothub.inbox_manager import InboxManager
+from azure.iot.device.iothub.aio.async_inbox import AsyncClientInbox
+
+pytestmark = pytest.mark.asyncio
+logging.basicConfig(level=logging.DEBUG)
+
+all_internal_handlers = [MESSAGE, METHOD, TWIN_DP_PATCH]
+all_handlers = [s.lstrip("_") for s in all_internal_handlers]
+
+
+@pytest.fixture
+def inbox_manager(mocker):
+    return InboxManager(inbox_type=AsyncClientInbox)
+
+
+@pytest.fixture
+def handler_manager(inbox_manager):
+    return AsyncHandlerManager(inbox_manager)
+
+
+# ----------------------
+# We have to do some unfortunate things here in order to manually mock out handlers, to test
+# tha they have been called. We can't use MagicMocks because not only do they not work well
```

Comment:
>MagicMocks [](start = 42, length = 10)

Not sure if this helps or not, but you might look at the unittest.mock.AsyncMock class.

Reply:
That could somewhat help, but doesn't really address the second point - these handlers are themselves the things under test, so mocking them out inherently takes away from the value of the test itself

---
In reply to: [456686506](https://github.com/Azure/azure-iot-sdk-python/pull/620#discussion_r456686506) [](ancestors = 456686506)